Exhibitor Support

yuusaku-t opened the pull request:
This PR is the same of #3357

This patch provides the integration of Druid & [Netflix Exhibitor](https://github.com/Netflix/exhibitor/wiki).

Exhibitor is a supervisor system for ZooKeeper. Exhibitor can dynamically scale-up/down the cluster of ZooKeeper servers. Druid currently use Apache Curator as ZooKeeper client. Curator can be integrated with Exhibitor to achieve a live/updating list of the ZooKeeper ensemble. For details, please refer to [this link](http://curator.apache.org/exhibitor.html).

I want to use Exhibitor-supervised Zookeeper cluster for Druid. Please check this pull request.

Thanks.

yuusaku-t commented:
@fjy @jon-wei I rewrote codes to use ```druid.zk.service.host``` as the backup connection string.
Please check this pull request.

Thanks.

drcrallen commented:
*glee*

I cannot express how glad I am proper usage of Exhibitor is not something I had to write, thank you!

I think falling back to `druid.zk.service.host` is the correct way to handle that setting.

I'm very interested in seeing this PR come through so it'll be high on my TODO list

drcrallen commented on server/src/main/java/io/druid/curator/CuratorModule.java:
```diff
@@ -54,20 +59,25 @@ public void configure(Binder binder)
-                               .build();
+        newCuratorBuilder(config, exConfig)
+            .sessionTimeoutMs(config.getZkSessionTimeoutMs())
+            .retryPolicy(new BoundedExponentialBackoffRetry(1000, 45000, 30))
```
Can you move these magic numbers to static constants at the top of the class?

Eventually they should be moved to their own config entries.

yuusaku-t replied:
OK. I'll try to rewrite codes to use static constants.

drcrallen commented on server/src/main/java/io/druid/curator/CuratorModule.java:
```diff
@@ -91,6 +101,54 @@ public void stop()
+    if (provider.isPresent()) {
+      return builder.ensembleProvider(provider.get());
+    }
+    return builder.connectString(config.getZkHosts());
```
under the hood this just does a ` new FixedEnsembleProvider(connectString)`, can `newEnsembleProvider` be modified to use `FixedEnsembleProvider` if no appropriate `ExhibitorConfig ` stuff is found

drcrallen commented on server/src/main/java/io/druid/curator/CuratorModule.java:
```diff
@@ -91,6 +101,54 @@ public void stop()
+      provider.pollForInitialEnsemble();
+      return Optional.<EnsembleProvider>of(provider);
+    }
+    catch (Exception err) {
```
It is not clear to me this should succeed. For example, if the polling is stuck and there's an `InterruptedException`, or a `RuntimeException` that wraps the interrupted information, this should fail.

Is there a reason the initial polling needs to take place?

yuusaku-t replied:
```ExhibitorEnsembleProvider#pollForInitialEnsemble()``` should be called before calling ```CuratorFramework#start()```;
otherwise, the curator client will fall back to ```druid.zk.service.host``` at initial connect time.
Because the curator client immediately after start does not have a list of ZooKeeper servers.
There is a certain time lag before curator client gets a list of ZooKeeper servers.
https://github.com/apache/curator/blob/apache-curator-2.11.0/curator-client/src/main/java/org/apache/curator/ensemble/exhibitor/ExhibitorEnsembleProvider.java#L128

I will move the place to write ```ExhibitorEnsembleProvider#pollForInitialEnsemble()``` into ```Lifecycle.Handler#start()```.

drcrallen commented on server/src/main/java/io/druid/curator/ExhibitorConfig.java:
```diff
@@ -0,0 +1,91 @@
+    return builder.addAll(Splitter.on(',')
+                                  .trimResults()
+                                  .omitEmptyStrings()
+                                  .split(hostsText))
```
indentation is weird here.

drcrallen commented on server/src/main/java/io/druid/curator/ExhibitorConfig.java:
```diff
@@ -0,0 +1,91 @@
+    if (hostsText == null) {
+      return builder.build();
+    }
+    return builder.addAll(Splitter.on(',')
```
Why not just have the property be a JSON array of strings?

fjy commented:
I'm 👍  given that exhibitor remains optional

yuusaku-t commented:
@drcrallen I rewrote patches based on your reviews. Please confirm.

The test coverage for this patch is still low. In particular, I have not verified the expected order of Lifecycle.Handler instances. To do this, I need to add a getter method to the Lifecycle class, or take internal variables by reflection. Or I might be able to test it with mock objects of Lifecycle.

Also, I modified JsonConfigTesterBase a bit to change the value format of `druid.exhibitor.service.hosts` from CSV to JSON array.

himanshug commented on server/src/test/java/io/druid/curator/CuratorModuleTest.java:
```diff
@@ -0,0 +1,120 @@
+  {
+    Properties props = new Properties();
+    props.put(curatorHostKey, "hostA");
+    props.put(exhibitorHostsKey, "[]");
```
this shouldn't be necessary, why do we need to explicitly set exhibitor hosts to empty array, it should be that by default by itself.

yuusaku-t replied:
I agree. The default value of ```druid.exhibitor.service.hosts``` is empty array. Therefore, it is not necessary to explicitly specify an empty array. In this test case, I would like to verify that ```druid.zk.service.host``` will be used if it is empty even if you specify ```druid.exhibitor.service.hosts```.

himanshug replied:
well, general usecase would be that user wouldn't be setting it at all, wouldn't that be a more realistic test case.

drcrallen replied:
both are reasonable test cases

himanshug commented on server/src/main/java/io/druid/curator/CuratorModule.java:
```diff
@@ -91,6 +104,58 @@ public void stop()
+      public void start() throws Exception
+      {
+        log.info("Poll the list of zookeeper servers for initial ensemble");
+        provider.pollForInitialEnsemble();
```
why this has to be done from lifecycle handler and not directly after creation of the provider ?

drcrallen replied:
https://github.com/druid-io/druid/pull/3664#discussion_r86807564 for conversation

himanshug commented:
I'm not too familiar with Exhibitor, but based on this patch my understanding is that, when exhibitor is configured.... curator client first goes to exhibitor instance to get list of zookeeper hosts and then afterwards it is talking directly to zookeeper instances.
so, druid.zk.service.* settings are still utilized..... right ?

yuusaku-t commented:
@himanshug Thank you for review!

> druid.zk.service.* settings are still utilized..... right ?

Yes, that's right. ```druid.exhibitor.service.``` setting items are only for periodically updating the list of Zookeeper servers. ```druid.zk.service.``` settings is used to connect to the actual zookeeper server.

drcrallen commented on docs/content/configuration/index.md:
```diff
@@ -75,6 +75,23 @@ The following path is used for service discovery. It is **not** affected by `dru
+|`druid.exhibitor.service.useSsl`|Boolean flag for whether or not to use https protocol.|`false`|
+|`druid.exhibitor.service.pollingMs`|How ofter to poll the exhibitors for the list|`10000`|
+
+Note that `druid.zk.service.host` is used as a backup in case an Exhibitor instance can't be contacted.
```
suggest adding `and therefore should still be set`

drcrallen commented on server/src/main/java/io/druid/curator/CuratorModule.java:
```diff
@@ -23,48 +23,61 @@
+
+  static final String EXHIBITOR_CONFIG_PREFIX = "druid.exhibitor.service";
+
+  private static final int baseSleepTimeMs = 1000;
```
general convention is all caps for static. so something like `DEFAULT_BASE_SLEEP_TIME_MS`

drcrallen commented on server/src/main/java/io/druid/curator/CuratorModule.java:
```diff
@@ -23,48 +23,61 @@
+
+  private static final int baseSleepTimeMs = 1000;
+
+  private static final int maxSleepTimeMs = 45000;
```
same for all caps

drcrallen commented on server/src/main/java/io/druid/curator/CuratorModule.java:
```diff
@@ -23,48 +23,61 @@
+
+  private static final int maxSleepTimeMs = 45000;
+
+  private static final int maxRetries = 30;
```
same with all caps

drcrallen commented on server/src/test/java/io/druid/guice/JsonConfigTesterBase.java:
```diff
@@ -125,8 +126,13 @@ public void setup() throws IllegalAccessException
+        Class<?> fieldType = field.getType();
+        if (String.class.isAssignableFrom(fieldType)) {
           propertyValues.put(propertyKey, UUID.randomUUID().toString());
+        } else if (Collection.class.isAssignableFrom(fieldType)) {
```
Added functionality needs tests.

yuusaku-t commented:
@himanshug @drcrallen I rewrote the patch again based on your review. Please confirm.

himanshug commented:
:+1: thanks @yuusaku-t

drcrallen commented:
FYI, https://issues.apache.org/jira/browse/CURATOR-382
